### PR TITLE
Introduce psr to interop container decorator for plugin managers

### PR DESCRIPTION
**Pull request by Xerkus**
Enables plugin managers to accept non-interop psr containers as creation
context and transparently wrap as interop for backward compatibility

Provide a narrative description of what you are trying to accomplish:

- [ ] Are you fixing a bug?
  - [ ] Detail how the bug is invoked currently.
  - [ ] Detail the original, incorrect behavior.
  - [ ] Detail the new, expected behavior.
  - [ ] Base your feature on the `master` branch, and submit against that branch.
  - [ ] Add a regression test that demonstrates the bug, and proves the fix.
  - [ ] Add a `CHANGELOG.md` entry for the fix.

- [x] Are you creating a new feature?
  - [x] Why is the new feature needed? What purpose does it serve?
    This PR enables plugin managers extending from abstract plugin manager to accept non-interop psr containers as creation context in a backward compatible manner. It is useful when alternative containers are used eg with zend-expressive.
    Decoration is necessary since FactoryInterface typehints against interop and won't work with pure PSR https://github.com/zendframework/zend-servicemanager/blob/0aea76faf40350852fd6570cb4c1eb60a18ac872/src/Factory/FactoryInterface.php#L10 https://github.com/zendframework/zend-servicemanager/blob/0aea76faf40350852fd6570cb4c1eb60a18ac872/src/Factory/FactoryInterface.php#L37
    This change is required since FactoryInterface typehints against interop container
  - [x] How will users use the new feature?
    Users now can pass non-interop psr containers as a constructor argument to plugin managers
  - [ ] Base your feature on the `develop` branch, and submit against that branch.
    Unable as `develop` branch have changes targeted at next major release
  - [x] Add only one feature per pull request; split multiple features over multiple pull requests
  - [x] Add tests for the new feature.
  - [ ] Add documentation for the new feature.
    Change is internal, no documentation is needed.
  - [x] Add a `CHANGELOG.md` entry for the new feature.

- [ ] Is this related to quality assurance?
  

- [ ] Is this related to documentation?

### CODE REVIEW

File: src/AbstractPluginManager.php [hunk 1]
```diff
@@ -47,11 +48,16 @@
...
+     * @param null|ConfigInterface|ContainerInterface|PsrContainerInterface $configInstanceOrParentLocator
      * @param array $config
      */
     public function __construct($configInstanceOrParentLocator = null, array $config = [])
     {
+        if ($configInstanceOrParentLocator instanceof PsrContainerInterface
+            && ! $configInstanceOrParentLocator instanceof ContainerInterface
+        ) {
+            $configInstanceOrParentLocator = new PsrContainerDecorator($configInstanceOrParentLocator);
+        }
```
Comment (weierophinney) [inline src/AbstractPluginManager.php]: zend-servicemanager requires container-interop/container-interop `^1.2`, and the `ContainerInterface` it ships _extends the PSR-11 `ContainerInterface`_. As such, you can simplify this tremendously by changing the check to test the container against `Psr\Container\ContainerInterface` instead. There's no reason to decorate in that instance.

File: src/PsrContainerDecorator.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
+<?php
+
```
Comment (webimpress) [inline src/PsrContainerDecorator.php]: remove empty line

File: src/PsrContainerDecorator.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * @link      http://github.com/zendframework/zend-servicemanager for the canonical source repository
```
Comment (webimpress) [inline src/PsrContainerDecorator.php]: usually we are using `@see` and https links please

File: src/PsrContainerDecorator.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * @copyright Copyright (c) 2005-2018 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (webimpress) [inline src/PsrContainerDecorator.php]: As new file, just 2018 year

File: src/PsrContainerDecorator.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
```
Comment (webimpress) [inline src/PsrContainerDecorator.php]: please use link to the LICENSE.md in the repository (see other repos for reference)

Comment (Xerkus) [reply src/PsrContainerDecorator.php]: I am not sure if i should change dockblock in a single file. I think it should be project wide PR targeting develop

File: src/PsrContainerDecorator.php (hunk 4 above)
Comment (Xerkus) [reply src/PsrContainerDecorator.php]: Are those copyrights on per project or per file basis? Per project would need whole range.

**Comment by Xerkus** [comment]
Those changes target 3.x only and will be undone before merge into develop for 4.0 btw.

File: src/PsrContainerDecorator.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * Finds an entry of the container by its identifier and returns it.
+     *
+     * @param string $id Identifier of the entry to look for.
+     *
+     * @throws NotFoundExceptionInterface  No entry was found for **this** identifier.
+     * @throws ContainerExceptionInterface Error while retrieving the entry.
+     *
+     * @return mixed Entry.
```
Comment (weierophinney) [inline src/PsrContainerDecorator.php]: This whole docblock can become:

```php
/**
 * {@inheritDoc}
 */
```

as it does not differ from the `ContainerInterface`.

(Same comment applies to the `has()` method.)

File: test/AbstractPluginManagerTest.php [hunk 7]
```diff
@@ -60,6 +63,32 @@ public function testInjectCreationContextInFactories()
...
+        $invokableFactory->method('__invoke')
+            ->will($this->returnArgument(0));
+
+        $config = [
+            'factories' => [
+                'creation context container' => $invokableFactory,
+            ],
+        ];
+
+        $container     = $this->getMockBuilder(PsrContainerInterface::class)
```
Comment (charma) [inline test/AbstractPluginManagerTest.php]:
```suggestion
        $container = $this->getMockBuilder(PsrContainerInterface::class)
```

Comment (webimpress) [reply test/AbstractPluginManagerTest.php]: @charma Please see: https://discourse.zendframework.com/t/poll-zend-coding-standard-2-aligning-statements-arrays-and-comment-tags/729

Comment (charma) [reply test/AbstractPluginManagerTest.php]: Thanks for the Info. I didnt know.

File: src/PsrContainerDecorator.php (hunk 4 above)
Comment (weierophinney) [reply src/PsrContainerDecorator.php]: Per file.

File: src/PsrContainerDecorator.php (hunk 5 above)
Comment (weierophinney) [reply src/PsrContainerDecorator.php]: My policy during review is:

- New files get the new format, listing the current year only in the date range.
- Existing files that have changed get updated to the new format, and have the date range from the previous docblock.

**Pull request merged**